types/src/validator_verifier.rs

```diff
@@ -14,72 +14,102 @@ pub enum VerifyError {
+        self.voting_power
+    }
+}
+
+/// Supports validation of signatures for known authors with individual voting powers. This struct
+/// can be used for all signature verification operations including block and network signature
+/// verification, respectively.
+#[derive(Clone)]
+pub struct ValidatorVerifier<PublicKey> {
+    address_to_validator_info: HashMap<AccountAddress, ValidatorInfo<PublicKey>>,
+    /// The minimum voting power required to achieve a quorum
+    quorum_voting_power: u64,
+    /// Total voting power of all validators (cached from address_to_validator_info)
+    total_voting_power: u64,
 }
 
 impl<PublicKey: VerifyingKey> ValidatorVerifier<PublicKey> {
-    /// Initialize with a map of author to public key and set quorum size to default (`2f + 1`) or
-    /// zero if `author_to_public_keys` is empty.
-    pub fn new(author_to_public_keys: HashMap<AccountAddress, PublicKey>) -> Self {
-        let quorum_size = if author_to_public_keys.is_empty() {
+    /// Initialize with a map of account address to validator info and set quorum size to
+    /// default (`2f + 1`) or zero if `address_to_validator_info` is empty.
+    pub fn new(
+        address_to_validator_info: HashMap<AccountAddress, ValidatorInfo<PublicKey>>,
+    ) -> Self {
+        let total_voting_power = address_to_validator_info
+            .values()
+            .fold(0, |sum, x| sum + x.voting_power);
+        let quorum_voting_power = if address_to_validator_info.is_empty() {
             0
         } else {
-            author_to_public_keys.len() * 2 / 3 + 1
+            total_voting_power * 2 / 3 + 1
         };
         ValidatorVerifier {
-            author_to_public_keys,
-            quorum_size,
+            address_to_validator_info,
+            quorum_voting_power,
+            total_voting_power,
         }
     }
 
-    /// Initializes a validator verifier with specified quorum size.
-    pub fn new_with_quorum_size(
-        author_to_public_keys: HashMap<AccountAddress, PublicKey>,
-        quorum_size: usize,
+    /// Initializes a validator verifier with a specified quorum voting power.
+    pub fn new_with_quorum_voting_power(
+        address_to_validator_info: HashMap<AccountAddress, ValidatorInfo<PublicKey>>,
+        quorum_voting_power: u64,
     ) -> Result<Self> {
+        let total_voting_power = address_to_validator_info
+            .values()
+            .fold(0, |sum, x| sum + x.voting_power);
         ensure!(
-            quorum_size <= author_to_public_keys.len(),
-            "Quorum size is greater than the number of authors: author_to_public_keys.len(): {}, \
+            quorum_voting_power <= total_voting_power,
+            "Quorum voting power is greater than the sum of all voting power of authors: {}, \
              quorum_size: {}.",
-            author_to_public_keys.len(),
-            quorum_size
+            quorum_voting_power,
+            total_voting_power
         );
```

Comment:
This is ripe for a MIRAI pre-condition (cc/ @hermanventer )

Reply:
Agreed. The caller treats an error result as a non recoverable error, so the ensure! may as well be a precondition!, which documents the true intent in a much clearer fashion.